nucypher/cli/characters/stake.py

```diff
@@ -34,63 +34,155 @@
-    @click.option('--allocation-filepath', help="Path to individual allocation file", type=EXISTING_READABLE_FILE)
-    @functools.wraps(func)
-    def wrapper(*args, **kwargs):
-        return func(*args, **kwargs)
-
-    return wrapper
-
-
-# Args (poa, registry_filepath, config_file, provider_uri, staking_address, hw_wallet,
-#       beneficiary_address, allocation_filepath, worker_address)
-def _worker_options(func):
-    @_stake_options
-    @click.option('--worker-address', help="Address to assign as an Ursula-Worker", type=EIP55_CHECKSUM_ADDRESS)
-    @functools.wraps(func)
-    def wrapper(*args, **kwargs):
-        return func(*args, **kwargs)
-
-    return wrapper
+option_value = click.option('--value', help="Token value of stake", type=click.INT)
+option_lock_periods = click.option('--lock-periods', help="Duration of stake in periods.", type=click.INT)
+option_worker_address = click.option('--worker-address', help="Address to assign as an Ursula-Worker", type=EIP55_CHECKSUM_ADDRESS)
+
+
+class StakeHolderConfigOptions:
+
+    __option_name__ = 'config_options'
+
+    def __init__(self, provider_uri, poa, light, registry_filepath):
+        self.provider_uri = provider_uri
+        self.poa = poa
+        self.light = light
+        self.registry_filepath = registry_filepath
+
+    def create_config(self, emitter, config_file):
+        try:
+            return StakeHolderConfiguration.from_configuration_file(
+                emitter=emitter,
+                filepath=config_file,
+                provider_uri=self.provider_uri,
+                poa=self.poa,
+                light=self.light,
+                sync=False,
+                registry_filepath=self.registry_filepath)
+
+        except FileNotFoundError:
+            return actions.handle_missing_configuration_file(
+                character_config_class=StakeHolderConfiguration,
+                config_file=config_file)
+
+    def generate_config(self, config_root):
+
+        assert self.provider_uri is not None # FIXME: add an error message
+
+        return StakeHolderConfiguration.generate(
+            config_root=config_root,
+            provider_uri=self.provider_uri,
+            poa=self.poa,
+            light=self.light,
+            sync=False,
+            registry_filepath=self.registry_filepath)
+
+
+group_config_options = group_options(
+    StakeHolderConfigOptions,
+    provider_uri=option_provider_uri(),
+    poa=option_poa,
+    light=option_light,
+    registry_filepath=option_registry_filepath,
+    )
+
+
+class StakeHolderBasicCharacterOptions:
+
+    __option_name__ = 'basic_character_options'
+
+    def __init__(self, config_options, staking_address):
+        self.config_options = config_options
+        self.staking_address = staking_address
+
+    def create_object(self, emitter, config_file, beneficiary_address=None, allocation_filepath=None):
```

Comment:
So the difference between `StakeHolderBasicCharacterOptions` and `StakeHolderCharacterOptions` is the stuff needed for a preallocation staker.

Unfortunately, the `beneficiary_address` and `allocation_filepath` still leak into `StakeHolderBasicCharacterOptions`, albeit as `None` parameters but still feels a little weird since most of the `create_object` function is made up of logic for preallocation staker. Without the preallocation staker functionality, this method seems to boil down to the following:

```python
def create_object(self, emitter, config_file):
    stakeholder_config = self.config_options.create_config(emitter, config_file)
    return stakeholder_config.produce(
            initial_address=self.staking_address) 
```
^ *(I think the `individual_allocation` parameter for `produce` defaults to `None` so not needed to be specified)*

Perhaps it makes sense to just have two separate `create_object` methods for the two types of Option classes `StakeHolderBasicCharacterOptions` and `StakeHolderCharacterOptions`.

Also, would it make sense to rename the old`StakeHolderBasicCharacterOptions` to `StakeHolderCharacterOptions` and then rename the old `StakeHolderCharacterOptions` to `PreallocationStakeHolderCharacterOptions` (or something similar)?

@cygnusv wdyt? You would know about the nuance of this code better than I would.

Reply:
Yes, I wasn't sure about the names because I don't understand the area well. `PreallocationStakeHolderCharacterOptions` sounds good to me.

> but still feels a little weird since most of the create_object function is made up of logic for preallocation staker.

Yes, that's a good point. I'll try to rearrange the logic there.

Reply:
Changed in the new revision, both the names and the logic placement.